### ISSUE TITLE
Processing of inotify queue overflow in sd-journal is not entirely reliable

### ISSUE THREAD

**Issue by msekletar** [systemd/systemd#8198]
### Submission type

 - Bug report

### systemd version the issue has been seen with

- upstream master

### Used distribution

- Fedora

### In case of bug report: Expected behaviour you didn't see

Journal client should notice that journald unlinked journal files that client has opened and it should close them also it should resync its in-memory state so it is consistent with the current filesystem content.

### In case of bug report: Unexpected behaviour you saw

- I had a client that opened journal and did nothing for some time while journald did the bunch of rotations + vacuuming. I had the system configured in a way that journald was deleting old files very aggressively and inotify queue size was set to a very low value, 5 events in my case.

- When client eventually called ```sd_journal_wait()``` it didn't properly close all journal files that were in the meantime unlinked by the server.

### In case of bug report: Steps to reproduce the problem

- https://gist.github.com/msekletar/8193b5966639b4ce571626ce14ea99bb

**Comment by msekletar** [comment]
So far I think the problem is following piece of code in ```add_any_file()```

```C
f = ordered_hashmap_get(j->files, path);
if (f) {
        /* Mark this file as seen in this generation. This is used to GC old files in
         * process_q_overflow() to detect journal files that are still and discern them from those who
         * are gone. */
        f->last_seen_generation = j->generation;
        return 0;
}
```

If we got ```IN_Q_OVERFLOW``` bit set in the inotify event mask we call ```process_q_overflow()```, which bumps the generation counter and calls ```reiterate_all_paths()```. Eventually, we end up in the above piece of code. If a queue has overflown before kernel enqueued remove event for the given file then we will still have that file open and corresponding JournalFile object present in our hashmap. Hence we find it in the hashmap because we do a path-based lookup and we reassign generation counter and just return. Back in ```process_q_overflow()``` we will skip processing of that file because it will have the generation counter set for the current generation. Problem is that in ```add_any_file()``` we were looking at a different file that happened to have the same name as the file we've opened in the past.

Long story short I think we need to beef up the logic in ```add_any_file()``` and also check if the file that we are considering to add is *actually the same* file we already added in the past. Just because the path is the same it doesn't mean it is the same file.

**Comment by msekletar** [comment]
@portante I think you will be interested in this

**Comment by poettering** [comment]
my proposed fix waiting in #8222. Please have a look!
